odml/property.py

```diff
@@ -507,6 +514,79 @@ def dependency_value(self, new_value):
             new_value = None
         self._dependency_value = new_value
 
+    @property
+    def val_cardinality(self):
+        """
+        The value cardinality of a Property. It defines how many values
+        are minimally required and how many values should be maximally
+        stored. Use 'values_set_cardinality' to set.
+        """
+        return self._val_cardinality
+
+    @val_cardinality.setter
+    def val_cardinality(self, new_value):
+        """
+        Sets the values cardinality of a Property.
+
+        The following cardinality cases are supported:
+        (n, n) - default, no restriction
+        (d, n) - minimally d entries, no maximum
+        (n, d) - maximally d entries, no minimum
+        (d, d) - minimally d entries, maximally d entries
+
+        Only positive integers are supported. 'None' is used to denote
+        no restrictions on a maximum or minimum.
+
+        :param new_value: Can be either 'None', a positive integer, which will set the maximum
+                          or an integer 2-tuple of the format '(min, max)'.
+        """
+        invalid_input = False
+
+        # Empty values reset the cardinality to None.
+        if not new_value or new_value == (None, None):
+            self._val_cardinality = None
+
+        # Providing a single integer sets the maximum value in a tuple.
+        elif isinstance(new_value, int) and new_value > 0:
+            self._val_cardinality = (None, new_value)
+
+        # Only integer 2-tuples of the format '(min, max)' are supported to set the cardinality
+        elif isinstance(new_value, tuple) and len(new_value) == 2:
+            v_min = new_value[0]
+            v_max = new_value[1]
+
+            min_int = isinstance(v_min, int) and v_min >= 0
+            max_int = isinstance(v_max, int) and v_max >= 0
+
+            if max_int and min_int and v_max > v_min:
+                self._val_cardinality = (v_min, v_max)
+
+            elif max_int and not v_min:
+                self._val_cardinality = (None, v_max)
+
+            elif min_int and not v_max:
+                self._val_cardinality = (v_min, None)
+
+            else:
+                invalid_input = True
+        else:
+            invalid_input = True
+
+        if invalid_input:
+            msg = "Can only assign single int or int-tuples of the format '(min, max)'"
+            raise ValueError(msg)
+
+    def values_set_cardinality(self, min_val=None, max_val=None):
```

Comment:
I am actually in two minds about the naming of this method the longer I think about it. We usually use a factual tone with the defining verb at the beginning while still stating what a method does. `values_cardinality` does not really tell what the method is supposed to be doing, but `values_set_cardinality` is switched around as well.

Maybe `set_values_cardinality` would be an ok compromise? It would at least be in line with how we named most of the other methods.

Reply:
I'm in favour of `set_values_cardinality`.  More straightforward and fits the pattern.  Also, the alternatively, `x_set_cardinality` I think can be confusing since the word `set` becomes ambiguous in the sense of a verb vs a noun: "set the cardinality" vs "the cardinality of the set".

I think using the phrase "set cardinality" in general (I think it appears in a docstring somewhere too) might be an issue.  I think you use it to mean "the cardinality given that it is set" but again it often sounds like you're saying "the cardinality of the set" (of values, etc).  I assume you don't mean the latter because we never refer to values or properties as sets in general.

Reply:
It is meant in the way of "I want to set the values cardinality".